go/channelling/turnservice_manager.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * Spreed WebRTC.
+ * Copyright (C) 2013-2016 struktur AG
+ *
+ * This file is part of Spreed WebRTC.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package channelling
+
+import (
+	"log"
+	"sync"
+	"time"
+
+	"github.com/strukturag/spreed-turnservicecli/turnservicecli"
+)
+
+type TURNServiceManager interface {
+	TurnDataCreator
+}
+
+type turnServiceManager struct {
+	sync.Mutex
+	pleaders map[uint64]Sender
+
+	uri         string
+	accessToken string
+	clientID    string
+	turnService *turnservicecli.TURNService
+}
+
+func NewTURNServiceManager(uri string, accessToken string, clientID string) TURNServiceManager {
+	turnService := turnservicecli.NewTURNService(uri, 0, nil)
+	mgr := &turnServiceManager{
+		uri:         uri,
+		accessToken: accessToken,
+		clientID:    clientID,
+
+		turnService: turnService,
+		pleaders:    make(map[uint64]Sender),
+	}
+
+	turnService.Open(accessToken, clientID, "")
+	turnService.BindOnCredentials(mgr.onCredentials)
+	log.Println("Fetching TURN credentials from service")
+	go func() {
+		//time.Sleep(10000 * time.Millisecond)
+		turnService.Autorefresh(true)
+	}()
+	// Wait a bit, to give TURN service some time to populate credentials, so
+	// we avoid to have send them as an update for fast reconnecting clients.
+	time.Sleep(500 * time.Millisecond)
```

Comment:
I would rather wait with a timeout of 500ms on an event, so if the credentials are received faster, it will continue immediately.